api/magic_links/routes.py

```diff
@@ -0,0 +1,159 @@
+        }
+        new_link_json.update({"token": create_token(new_link_json)})
+        return new_link_json
+
+    def _set_unique_keyed_record(self, value) -> str | None:
+        """
+        Attempts to save the link claims with a unique short key
+        that doesn't already exist (and returns None if it fails)
+        :param value: A json dictionary of link attributes
+        :return: The unique short key for the link hash (or None)
+        """
+        key = self._make_short_key()
+        created = self.redis_mlinks.setex(
+            key, env.config.get("MAGIC_LINK_EXPIRY_SECONDS", 0), value
+        )
+        if created == 1:
+            return key
+
+    def search(self):
+        """
+        GET /magic-links endpoint
+        :return: Json Response
+        """
+        return Response(json.dumps(self._links), mimetype="application/json")
+
+    def use(self, link_id: str):
+        """
+        GET /magic-links/{link_id} endpoint
+        If the link_id matches a valid link key, this then:
+        - creates a session,
+        - sets a session_id cookie in the client
+        - sets a session_token cookie in the client
+        - deletes the link record from redis
+        - then finally, redirects to the redirect_url
+        If no matching valid link is found returns a 404 error message
+        :param link_id: String short key for the link
+        :return: 302 Redirect / 404 Error
+        """
+        link_hash = self.redis_mlinks.get(link_id)
+        if link_hash:
+            link = json.loads(link_hash)
+            self.redis_mlinks.delete(link_id)
+
+            # Check link is not expired
+            if link.get("exp") > int(datetime.now().timestamp()):
+                return AuthSessionView.create_session_and_redirect(
+                    account_id=link.get("accountId"),
+                    redirect_url=link.get("redirectUrl"),
+                )
+            return error_response(404, "Link expired")
+        return error_response(404, "Link expired or invalid")
+
+    def create(self, email: str, redirect_url: str = "https://google.com"):
+        """
+        Creates a magic link for an existing account holder
+        :param email: the account holders email address
+        :param redirect_url: the url the link should redirect to
+        :return: a json of the magic link created (or an error of failure)
+        """
+        account = get_account(email)
+        if account:
+            new_link_json = self._make_link_json(account, redirect_url)
+
+            # Try up to 6 times to create a unique short key for the link
+            # which doesn't already exist as a key in redis
+
+            tries = 0
+            link_key = None
+            while tries < 6:
+                link_key = self._set_unique_keyed_record(
+                    json.dumps(new_link_json)
+                )
+                if link_key:
+                    break
+                tries += 1
```

Comment:
If this retry logic is definitely required, perhaps we could wrap it in a helper function so it could be used by other Redis calls?

Reply:
There is a nicer way of doing this:
`for _ in range(6)
 if thing:
  break
:`

Reply:
Nice, done